### PR TITLE
Feature (Gateway): Ve-Boba Lock for Vote page.

### PR DESCRIPTION
**Pull request by sk-enya**
:clipboard: Closes, https://github.com/bobanetwork/ve-boba/issues/12

## Overview

Lock for vote page for ve boba

## Changes
- Added new page for lock
- Way to represent earlier, existing locks for the user
- Way to create new Lock
- Manage lock expired can be withdraw, other can have option to increase amount and extend time.
- Representation of voting power vs month in chart.

## Testing
- Navigate to lock page from Menu Items at top.
- Create new lock
- List of records or lock will be available as soon as lock has been created
- Pagination should work.
- Can have option to withdraw incase of expired lock.
- Can have option to increase amount or extend the expiry for existing lock which is yet to expired.
- Voting power and Voting ration will be calculated base on the selected expiry.
- Max expiry we have is 1yr.

### CODE REVIEW

File: packages/boba/gateway/src/components/areaChart/AreaChart.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import { Typography } from '@mui/material';
+import React from 'react';
+import { AreaChart as ReAreaChart, Area, Tooltip, XAxis, YAxis, ResponsiveContainer, CartesianGrid } from 'recharts';
+
+
+import * as S from './AreaChart.styles'
+
+/*
+TODO:
+  - make the chart label to be correct on X-Axis and Y-Axis
```
Comment (InoMurko) [inline packages/boba/gateway/src/components/areaChart/AreaChart.js]: Should we make these TODOs into github issues rather? They get lost in the code.

File: packages/boba/gateway/src/components/mainMenu/menuItems.js [hunk 2]
```diff
@@ -29,6 +29,12 @@ export const menuItems = [
     title: "Earn",
     url: "/earn",
   },
+  // {
```
Comment (InoMurko) [inline packages/boba/gateway/src/components/mainMenu/menuItems.js]: Can we manage this ve feature as ENV vars?

The step to production is to enable it on a testnet, like Rinkeby, test it there and having the option to enable it. Right @souradeep-das ?

File: packages/boba/gateway/src/components/areaChart/AreaChart.js (hunk 1 above)
Comment (sk-enya) [reply packages/boba/gateway/src/components/areaChart/AreaChart.js]: opps missed to remove the comment, it was for me to just remember as I did while doing brainstorming.

File: packages/boba/gateway/src/components/mainMenu/menuItems.js (hunk 2 above)
Comment (sk-enya) [reply packages/boba/gateway/src/components/mainMenu/menuItems.js]: hmm sure we can manage it with ENV let me get it added.

Comment (sk-enya) [reply packages/boba/gateway/src/components/mainMenu/menuItems.js]: done.

File: packages/boba/gateway/src/containers/veboba/Lock.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import CreateLock from './createLock/CreateLock'
+import LockRecords from './Records/Records'
+import { useDispatch, useSelector } from 'react-redux'
+import { selectAccountEnabled } from 'selectors/setupSelector'
+
+const data = [
+  { name: '0', uv: 0 },
+  { name: '3M', uv: 25 },
+  { name: '6M', uv: 50 },
+  { name: '9M', uv: 75 },
```
Comment (InoMurko) [inline packages/boba/gateway/src/containers/veboba/Lock.js]: Below you have expiry options 3,6,12. Is 9 an option?

File: packages/boba/gateway/src/containers/veboba/createLock/CreateLock.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+import { openAlert } from 'actions/uiAction'
+
+import * as S from './CreateLock.styles'
+import * as Styles from './CreateLock.module.scss'
+import "react-datepicker/dist/react-datepicker.css"
+import { useRef } from 'react'
+import { selectLoading } from 'selectors/loadingSelector'
+
+
+const EXPIRY_OPTIONS = [
```
Comment (InoMurko) [inline packages/boba/gateway/src/containers/veboba/createLock/CreateLock.js]: Here

File: packages/boba/gateway/src/reducers/veBobaReducer.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+/*
+Copyright 2019-present OmiseGO Pte Ltd
```
Comment (InoMurko) [inline packages/boba/gateway/src/reducers/veBobaReducer.js]: Boba Network :) please check if it's anywhere elsewhere too

File: packages/boba/gateway/src/services/networkService.js [hunk 6]
```diff
@@ -4621,6 +4628,238 @@ class NetworkService {
     return bobaBridges.filter((bridge) => bridge.tokens.includes(tokenSymbol))
   }
 
+  /***********************************************/
```
Comment (InoMurko) [inline packages/boba/gateway/src/services/networkService.js:4631]: NetworkService file is becoming one huge spaghetti... are there better ways to refactor this? (Not in this PR, but if you too think this is a problem, please create a GH issue)

**Comment by InoMurko** [comment]
since you added a new REACT_APP_... env var, make sure it's documented where all other vars are

File: packages/boba/gateway/src/containers/veboba/Lock.js (hunk 3 above)
Comment (sk-enya) [reply packages/boba/gateway/src/containers/veboba/Lock.js]: Hmm, but in that case graph won't look linear.

Comment (InoMurko) [reply packages/boba/gateway/src/containers/veboba/Lock.js]: Is 9 a possible lock option @souradeep-das ?

File: packages/boba/gateway/src/services/networkService.js (hunk 6 above)
Comment (sk-enya) [reply packages/boba/gateway/src/services/networkService.js:4631]: Yeah, we can there are multiple approaches.yeah it's going to be problem..

Created issues.
https://github.com/bobanetwork/boba/issues/242

**Comment by sk-enya** [comment]
> since you added a new REACT_APP_... env var, make sure it's documented where all other vars are

As of now I didn't saw any documentation for the env for `REACT_APP_....` I will add that details into the readme of gateway.

**Comment by InoMurko** [comment]
> > since you added a new REACT_APP_... env var, make sure it's documented where all other vars are
> 
> 
> 
> As of now I didn't saw any documentation for the env for `REACT_APP_....` I will add the it's details into the readme it self.

Thank you

File: packages/boba/gateway/src/reducers/veBobaReducer.js (hunk 5 above)
Comment (sk-enya) [reply packages/boba/gateway/src/reducers/veBobaReducer.js]: This is at multiple places will create another PR to remove it later on.

Comment (sk-enya) [reply packages/boba/gateway/src/reducers/veBobaReducer.js]: https://github.com/bobanetwork/boba/issues/243 created issue will raise another pr for other files.

File: packages/boba/gateway/src/containers/veboba/Lock.js (hunk 3 above)
Comment (sk-enya) [reply packages/boba/gateway/src/containers/veboba/Lock.js]: Only showing to 3,6 and 1yr now

**Review by InoMurko — APPROVED** [review]

**Comment by boyuan-chen** [comment]
> > since you added a new REACT_APP_... env var, make sure it's documented where all other vars are
> 
> As of now I didn't saw any documentation for the env for `REACT_APP_....` I will add that details into the readme of gateway.

Could you please also add it to `.env.example`?

**Comment by boyuan-chen** [comment]
The default input is `0`. Probably use the placeholder as `0`?
image

**Comment by boyuan-chen** [comment]
I tried to lock more than one year and got this `Internal JSON-RPC error.`, but I didn't know what the actual error is. I checked the console and the actual error is `execution reverted: Voting lock can be 1 year max`.

Could you please disable the button when the time range is more than 1 year?

**Comment by boyuan-chen** [comment]
I forgot to enter the BOBA balance and picked 6 months and got this error when I clicked the `Lock` button. 
```
invalid BigNumber string (argument="value", value="NaN", code=INVALID_ARGUMENT, version=bignumber/5.5.0)
```

Probably disable the button when the input is `null`?

**Comment by boyuan-chen** [comment]
I click `Extend Lock Time` and didn't change the proposed time above and got the `Internal JSON-RPC error`. However, I didn't know the actually error and I found 
```
execution reverted: Can only increase lock duration
```
It turns out that I have to change the time?
image

**Comment by sk-enya** [comment]
> Could you please also add it to `.env.example`?

Yeah it's already added!
https://github.com/bobanetwork/boba/pull/222/files#diff-5220f01182126b5bcb08dbcce1a287a275416d24105329b3ddb5525f1ab8d266

**Comment by sk-enya** [comment]
> Probably disable the button when the input is `null`?

fixed!

File: packages/boba/gateway/src/components/areaChart/AreaChart.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  label
+}) {
+
+  if (!active) {
+    return null;
+  }
+
+  return <S.ToolTipContainer borderRadius={2} p={2}>
+    <Typography variant='body2'>Locking period: {label} </Typography>
+    <Typography variant='body2'>Convert ration: {payload[ 0 ].value}</Typography>
```
Comment (souradeep-das) [inline packages/boba/gateway/src/components/areaChart/AreaChart.js]: minor: typo ration -> ratio

File: packages/boba/gateway/src/containers/veboba/Lock.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+
+  return <S.PageContainer>
+    <Box display="flex" container gap={3} flexDirection="column">
+      <Box display="flex" justifyContent='space-around' gap={4}>
+        <Box sx={{ width: '100%' }}>
+          <Typography variant="h2">Lock for Vote</Typography>
+          <Typography variant="body2" sx={{ opacity: 0.6 }}>More tokens locked for longer = greater voting power = higher rewards</Typography>
+          <G.DividerLine sx={{ my: 2 }} />
+          <Typography variant="h3">Locking Period vs Convert Ration</Typography>
```
Comment (souradeep-das) [inline packages/boba/gateway/src/containers/veboba/Lock.js]: minor: typo Ration -> Ratio

**Comment by souradeep-das** [comment]
looks great! but here are some things I found:

1/ locks which expire even in one week are shown "expires in a month". Could we replace this to something like "expires this month" or expires in a week"
Screenshot 2022-07-15 at 3 33 31 PM

2/ the minimum lock period is 1 week, can the calender grey out days until July 22 here?
Screenshot 2022-07-15 at 3 35 33 PM

3/ Extending lock time - updates the text below "Your voting power will be x -> x' "
but increasing the amount does not, could we do that for increase amount as well?
Screenshot 2022-07-15 at 3 42 00 PM

**Comment by sk-enya** [comment]
@souradeep-das 
1. fix pushed it  will be looks something like 
Screenshot 2022-07-18 at 9 36 39 AM
Screenshot 2022-07-18 at 9 36 43 AM

2. fix pushed
3. Good catch, Yes, on increasing the amount added the fix to update voting power.

File: packages/boba/gateway/src/containers/veboba/Records/RecordItem.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+  const sameMonth = moment(expiry).isSame(moment(), 'month');
+  const sameWeek = moment(expiry).isSame(moment(), 'week');
+
+  let expiryText = '';
+  if (sameMonth) {
+    expiryText= 'Expires this month'
+  }
+  if (sameWeek) {
+    expiryText= 'Expires in week'
```
Comment (souradeep-das) [inline packages/boba/gateway/src/containers/veboba/Records/RecordItem.js]:
```suggestion
    expiryText= 'Expires in a week'
```

**Review by souradeep-das — APPROVED** [review]
amazing!

**Pull request merged**